### PR TITLE
Add remainder of FMCAD23 partitioning work

### PR DESCRIPTION
**Pull request by amaleewilson**
This adds the remainder of the FMCAD partitioning work and also fixes a bug that was introduced in the first PR (not a bug in the original code, but it was an unintended side effect of refactoring).

### CODE REVIEW

File: src/theory/theory_engine.cpp
Comment (ajreynol) [inline src/theory/theory_engine.cpp:1454]: This should be part of `PartitionGenerator::notifyLemma`.

On a related note, it seems that the check on line 1463 was previously wrong (I had fixed this on my branch https://github.com/ajreynol/CVC4/tree/plugin, but this was not merged to main).

Can you change the condition `if (d_relManager != nullptr)` to `if (!d_modules.empty())`?

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/partition_generator.cpp
Comment (amaleewilson) [inline src/theory/partition_generator.cpp:362]: Oops, this long comment got collapsed into nonsense with the auto-formatting. Will fix it.

File: src/theory/theory_engine.cpp
Comment (amaleewilson) [reply src/theory/theory_engine.cpp]: Fixed both.

**Comment by amaleewilson** [comment]
Marking this as do not merge for now - if there are a lot of changes requested, I want to make sure that this is thoroughly tested against the old branch before getting merged.

File: src/theory/partition_generator.cpp
Comment (ajreynol) [inline src/theory/partition_generator.cpp:125]: Shouldn't you just merge these sets into one?

File: src/theory/partition_generator.cpp
Comment (ajreynol) [inline src/theory/partition_generator.cpp:131]: You probably don't need the second check here, the check on line 138 is sufficient.

File: src/theory/partition_generator.cpp
Comment (ajreynol) [inline src/theory/partition_generator.cpp:138]: This could just be one call if you merge the sets of unusable kinds into one set.

File: src/theory/partition_generator.h
Comment (ajreynol) [inline src/theory/partition_generator.h:231]: Nit: maybe use unsigned int here.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/partition_generator.cpp
Comment (amaleewilson) [reply src/theory/partition_generator.cpp:138]: Good point, thank you!

**Comment by amaleewilson** [comment]
Removing the do not merge label because I have confirmed we see expected results with all of the partitioning strategies. Will rerun tests before committing any further changes.

**Review by ajreynol — APPROVED** [review]
LGTM, thanks for the changes.

**Pull request merged**
